Memory

hjorthjort opened the pull request:
This PR implements memory load and store operations

hjorthjort commented:
@ehildenb This PR is almost done, but I'm having some issues I really can't figure out, with the folded syntax. Have a look at the `memory.wast` test file, at the bottom.

ehildenb commented on data.md:
```diff
@@ -6,6 +6,8 @@ require "domains.k"
 module WASM-DATA
     imports DOMAINS
+    imports BYTES
+//    imports COLLECTIONS
```
lets not include things that we comment out

ehildenb commented on data.md:
```diff
@@ -83,10 +85,13 @@ We can append two `ValTypes`s with the `_+_` operator.
-    syntax Int ::= #width ( IValType ) [function]
+    syntax Int ::= #width    ( IValType ) [function]
+    syntax Int ::= #numBytes ( IValType ) [function]
  // ---------------------------------------------
```
formatting

hjorthjort commented on tests/simple/memory.wast:
```diff
@@ -32,5 +32,95 @@
+(i32.const 1)
+(i32.load8_s)              ;; Doesn't work!
+#assertTopStack <i32> 0 "load8 signed"
+(i32.load8_s (i32.const 1))              ;; Doesn't work!
```
This doesn't work. For some reason, the interpreter fails to unify `( i32 . load8_s ( i32 . const 1 ) ) ` with the rule on line 758 in `wasm.md`.

hjorthjort commented on wasm.md:
```diff
@@ -684,10 +688,133 @@ Currently, only one memory may be accessible to a module, and thus the `<memAddr
+                   | IValType "." LoadOp Int
+                   | "load" "{" IValType Int Int Signedness "}"
+ // -----------------------------------------------------------
+    rule <k> ( ITYPE . LOP:LoadOp MEMARG:MemArg I:Instr ) => I ~> ( ITYPE . LOP MEMARG ) ... </k>
```
It seems this rule should be applied in the offending test case, but it isn't.

ehildenb commented on tests/simple/memory.wast.draft:
```diff
@@ -0,0 +1,58 @@
+  (memory 1)
+  (data (i32.const 0) "WASM")
+  (i32.load8_u (i32.const 2))
+  #assertTopStack < i32 > 83 "S in bytes" 
```
Probably shouldn't have `#assert_` statements directly inside modules, as the semantics of modules will eventually not be "execute everything in the module like it was inline code anyway".

Also, we should be able to test `memory` assuming no modules exist yet. So if you want integration tests with `module`, I'd put these tests in `modules.wast`. If they are testing actual features of memories (and not just testing integration with modules), can we make them standalone memory tests instead?

hjorthjort replied:
Sorry, this is from the WIP I integrated, this file is not supposed to be included.

MrChico commented on data.md:
```diff
@@ -200,5 +216,58 @@ Operator `_++_` implements an append operator for sort `Stack`.
+        // Don't store 0 bytes.
+        #if VAL modInt 256 =/=Int 0
+          #then BM [IDX <- VAL modInt 256 ]
+          #else BM [IDX <- undef          ]
```
Hmm, I think splitting up these as different rules rather than splitting on the right hand side would be nicer

ehildenb replied:
The Java K prover will only apply a given function rule if it can show that we are fully subsumed by that case. If you promote the splitting to the top-level rule (as @MrChico suggests), it will avoid duplicating the existing map `BM` twice in memory if it cannot prove that only one case of the function applies.

Another option is to fold the `#if_#then_#else_#fi` into the update itself:

```
...   => BM [ IDX <- #if VAL modInt 256 =/=Int 0 #then VAL modInt 256 #else undef ] ...
```

This avoids the memory explosion that you'll get with duplicating `BM` on the RHS when you can't prove that either `VAL modInt 256 =/=Int 0` or the negation is true, so would be another solution to the problem.

But I think @MrChico's solution is the best, because the term `BM [ IDX := VAL ]` held in memory is much smaller and easier to write lemmas about than the case where we evaluate the map fully.

hjorthjort replied:
Thanks for the explanation, done.

ehildenb commented on data.md:
```diff
@@ -200,5 +216,58 @@ Operator `_++_` implements an append operator for sort `Stack`.
+ // ----------------------------------------------------
+    rule #range(BM:Map, START, 0    ) => 0
+    rule #range(BM:Map, START, WIDTH) => #lookup(BM, START) +Int (#range(BM, START +Int 1, WIDTH -Int 1) *Int 256)
+      requires WIDTH >Int 0 [concrete]
```
The Java prover will have a better time if the first rule is phrased as:

```
rule #range(BM, START, WIDTH) => 0 requires WIDTH ==Int 0
```

The reason is that a match on the LHS of a rule will be translated into `==K`, but here we can use the more specific `==Int` (which has it's own reasoning builtin).

Another option would be to mark *both* rules as `concrete`.

hjorthjort replied:
Great explanation, thanks. It still makes sense to mark the second rule concrete anyway, correct?

ehildenb commented on data.md:
```diff
@@ -200,5 +216,58 @@ Operator `_++_` implements an append operator for sort `Stack`.
+    syntax Map ::= #clearRange(Map, Int, Int) [function]
+ // ----------------------------------------------------
+    rule #clearRange(M, START, END) => M                                     requires START >=Int END
+    rule #clearRange(M, START, END) => #clearRange(M [START <- undef], START +Int 1, END) requires START <Int END
```
formatting (line up `requires`, only use spaces)

hjorthjort replied:
Done (used linebreak)

ehildenb commented on wasm.md:
```diff
@@ -684,12 +688,131 @@ Currently, only one memory may be accessible to a module, and thus the `<memAddr
+      < ITYPE > #if SIGN ==K Signed
+                #then #signedWidth(WIDTH, #range(DATA, EA, WIDTH /Int 8))
+                #else #range(DATA, EA, WIDTH /Int 8)
+                #fi ... </k>
```
Formatting recommendation: 

```
    rule <k> load { ITYPE WIDTH EA SIGN }
          => < ITYPE > #if SIGN ==K Signed
                           #then #signedWidth(WIDTH, #range(DATA, EA, WIDTH /Int 8))
                           #else #range(DATA, EA, WIDTH /Int 8)
                       #fi 
         ...
         </k>
```

hjorthjort replied:
Good, thanks.

ehildenb commented on wasm.md:
```diff
@@ -684,12 +688,131 @@ Currently, only one memory may be accessible to a module, and thus the `<memAddr
+           <mdata>   DATA </mdata>
+           ...
+         </memInst>
+         requires (EA +Int WIDTH /Int 8) <=Int (SIZE *Int #pageSize())
```
De-indent `requires` to make it more visually apparent (usually I line up the `r` in `requires` with the `l` in `rule`).

hjorthjort replied:
Done

hjorthjort commented on data.md:
````diff
@@ -200,5 +216,58 @@ Operator `_++_` implements an append operator for sort `Stack`.
+ // -----------------------------------------------
+    rule #lookup( (KEY |-> VAL) M, KEY ) => VAL                               [concrete]
+    rule #lookup(               M, KEY ) => 0 requires notBool KEY in_keys(M) [concrete]
+```
````
Note to self: you can probably use the `orDefault` construction in the standard library.

hjorthjort commented:
I addressed your comments, thanks both.
I also fixed that small bug in `memory.grow`. I figured that's alright in a big memory PR, not worth opening a separate PR for.
